File: apps/docker-server/cmd/renewlet/subscription_list_filters.go
package main

// subscription_list_filters.go 只在 owner-scoped 派生投影上筛选和分页，再一次性回表读取完整事实记录。
// cursor 只裁当前页，total 始终来自同一完整过滤集，避免滚动后总数递减或筛选口径漂移。

import (
	"errors"
	"fmt"
	"net/url"
	"strconv"
	"strings"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

const (
	subscriptionListDefaultLimit       = 50
	subscriptionListMaxLimit           = 100
	subscriptionListScanPageSize       = 500
	subscriptionListSearchMaxLength    = 200
	subscriptionPaymentMethodNoneValue = "__none"
)

type subscriptionListQuery struct {
	Limit           int
	Cursor          *subscriptionCursorPayload
	Search          string
	Categories      []string
	Tags            []string
	BillingCycles   []string
	PaymentMethods  []string
	Currencies      []string
	Status          string
	Renewal         string
	NextBillingFrom string
	NextBillingTo   string
	Pinned          *bool
	PublicHidden    *bool
	ReminderMode    string
	RepeatReminder  *bool
}

type subscriptionListPage struct {
	Rows       []*core.Record
	NextCursor *string
	Total      int64
}

func parseSubscriptionListQuery(values url.Values) (subscriptionListQuery, error) {
	limit, err := parsePositiveQueryInt(values.Get("limit"), subscriptionListDefaultLimit, 1, subscriptionListMaxLimit)
	if err != nil {
		return subscriptionListQuery{}, err
	}
	query := subscriptionListQuery{Limit: limit}
	if rawCursor := strings.TrimSpace(values.Get("cursor")); rawCursor != "" {
		cursor, err := parseSubscriptionCursorPayload(rawCursor)
		if err != nil {
			return subscriptionListQuery{}, err
		}
		query.Cursor = &cursor
	}
	if search := strings.TrimSpace(values.Get("q")); search != "" {
		if len(search) > subscriptionListSearchMaxLength {
			return subscriptionListQuery{}, errors.New("invalid search query")
		}
		query.Search = search
	}
	if query.Categories, err = parseSubscriptionListStrings(values["category"], 50, 80, nil); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.Tags, err = parseSubscriptionListStrings(values["tag"], 100, 40, nil); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.BillingCycles, err = parseSubscriptionListStrings(values["billingCycle"], 7, 40, isValidBillingCycle); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.PaymentMethods, err = parseSubscriptionListStrings(values["paymentMethod"], 200, 80, isValidSubscriptionListPaymentMethod); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.Currencies, err = parseSubscriptionListStrings(values["currency"], 50, 3, isSubscriptionListCurrency); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.Status, err = parseSubscriptionListSingle(values, "status", 40, isValidSubscriptionStatus); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.Renewal, err = parseSubscriptionListSingle(values, "renewal", 20, isSubscriptionListRenewal); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.NextBillingFrom, err = parseSubscriptionListSingle(values, "nextBillingFrom", 10, isValidDateOnly); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.NextBillingTo, err = parseSubscriptionListSingle(values, "nextBillingTo", 10, isValidDateOnly); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.NextBillingFrom != "" && query.NextBillingTo != "" && query.NextBillingFrom > query.NextBillingTo {
		return subscriptionListQuery{}, errors.New("invalid next billing range")
	}
	if query.Pinned, err = parseSubscriptionListBool(values, "pinned"); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.PublicHidden, err = parseSubscriptionListBool(values, "publicHidden"); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.ReminderMode, err = parseSubscriptionListSingle(values, "reminderMode", 20, isSubscriptionListReminderMode); err != nil {
		return subscriptionListQuery{}, err
	}
	if query.RepeatReminder, err = parseSubscriptionListBool(values, "repeatReminder"); err != nil {
		return subscriptionListQuery{}, err
	}
	return query, nil
}

func listSubscriptionRecordsForQuery(app core.App, userID string, query subscriptionListQuery, today string) (subscriptionListPage, error) {
	// 该入口的查询预算固定为一次投影页查询和一次批量事实回表；禁止恢复逐 ID PocketBase 查询。
	pageIDs, total, err := projectedSubscriptionPageIDs(app, userID, query, today)
	if err != nil {
		return subscriptionListPage{}, err
	}
	rows, err := getSubscriptionRecordsByIDs(app, userID, pageIDs)
	if err != nil {
		return subscriptionListPage{}, err
	}
	var nextCursor *string
	if len(rows) > query.Limit {
		rows = rows[:query.Limit]
		cursor := encodeSubscriptionCursor(rows[len(rows)-1])
		nextCursor = &cursor
	}
	return subscriptionListPage{Rows: rows, NextCursor: nextCursor, Total: total}, nil
}

func projectedSubscriptionPageIDs(app core.App, userID string, query subscriptionListQuery, today string) ([]string, int64, error) {
	base := subscriptionProjectionBaseQuery(userID, query)
	if query.Search != "" {
		base.conditions = append(base.conditions, "instr(idx.search_text_lower, {:search}) > 0")
		base.params["search"] = strings.ToLower(strings.TrimSpace(query.Search))
	}
	if query.Status != "" {
		base.conditions = append(base.conditions, `(CASE
			WHEN idx.status = 'expired' THEN 'expired'
			WHEN idx.billing_cycle = 'one-time' AND idx.one_time_term_count <= 0 THEN idx.status
			WHEN idx.status IN ('active', 'trial') AND idx.next_billing_date < {:today} THEN 'expired'
			ELSE idx.status
		END) = {:status}`)
		base.params["today"] = today
		base.params["status"] = query.Status
	}
	rows, err := runSubscriptionProjectionPage(app, base, query.Limit+1, query.Cursor)
	if err != nil {
		return nil, 0, err
	}
	if len(rows) == 0 {
		return []string{}, 0, nil
	}
	return subscriptionProjectionIDs(rows), int64(rows[0].TotalCount), nil
}

type subscriptionProjectionBase struct {
	conditions []string
	params     dbx.Params
}

func subscriptionProjectionBaseQuery(userID string, query subscriptionListQuery) subscriptionProjectionBase {
	base := subscriptionProjectionBase{
		conditions: []string{"idx.user_id = {:user}"},
		params:     dbx.Params{"user": userID},
	}
	appendSQLInCondition(&base, "idx.category", "category", query.Categories)
	appendSQLInCondition(&base, "idx.billing_cycle", "billingCycle", query.BillingCycles)
	appendSQLInCondition(&base, "idx.currency", "currency", query.Currencies)
	appendSQLPaymentMethodCondition(&base, query.PaymentMethods)
	appendSQLRenewalCondition(&base, query.Renewal)
	appendSQLTagCondition(&base, query.Tags)
	if query.NextBillingFrom != "" {
		base.conditions = append(base.conditions, "idx.next_billing_date >= {:nextBillingFrom}")
		base.params["nextBillingFrom"] = query.NextBillingFrom
	}
	if query.NextBillingTo != "" {
		base.conditions = append(base.conditions, "idx.next_billing_date <= {:nextBillingTo}")
		base.params["nextBillingTo"] = query.NextBillingTo
	}
	if query.Pinned != nil {
		base.conditions = append(base.conditions, "idx.pinned = {:pinned}")
		base.params["pinned"] = boolToSQLiteInt(*query.Pinned)
	}
	if query.PublicHidden != nil {
		base.conditions = append(base.conditions, "idx.public_hidden = {:publicHidden}")
		base.params["publicHidden"] = boolToSQLiteInt(*query.PublicHidden)
	}
	appendSQLReminderModeCondition(&base, query.ReminderMode)
	if query.RepeatReminder != nil {
		base.conditions = append(base.conditions, "idx.repeat_reminder_enabled = {:repeatReminder}")
		base.params["repeatReminder"] = boolToSQLiteInt(*query.RepeatReminder)
	}
	return base
}

func runSubscriptionProjectionPage(app core.App, base subscriptionProjectionBase, limit int, cursor *subscriptionCursorPayload) ([]subscriptionListIndexRow, error) {
	pageConditions := []string{"1 = 1"}
	params := dbx.Params{}
	for key, value := range base.params {
		params[key] = value
	}
	if cursor != nil {
		pageConditions = append(pageConditions, "(idx.created_at < {:cursorCreatedAt} OR (idx.created_at = {:cursorCreatedAt} AND idx.subscription_id < {:cursorID}))")
		params["cursorCreatedAt"] = cursor.CreatedAt
		params["cursorID"] = cursor.ID
	}
	params["limit"] = limit
	var rows []subscriptionListIndexRow
	// filtered CTE 先锁定 owner 并计算完整过滤集；外层 cursor 只裁当前页，不改变 total。
	err := app.DB().NewQuery(fmt.Sprintf(`WITH filtered AS (
			SELECT subscription_id, user_id, name, website, notes, search_text_lower, category, billing_cycle, currency,
				payment_method, status, pinned, public_hidden, next_billing_date, trial_end_date, one_time_term_count,
				auto_renew, reminder_days, repeat_reminder_enabled, created_at, updated_at
			FROM subscription_list_index AS idx
			WHERE %s
		), page AS (
			SELECT * FROM filtered AS idx
			WHERE %s
			ORDER BY idx.created_at DESC, idx.subscription_id DESC
			LIMIT {:limit}
		), totals AS (
			SELECT COUNT(*) AS total_count FROM filtered
		)
			SELECT COALESCE(page.subscription_id, '') AS subscription_id, COALESCE(page.user_id, '') AS user_id,
			COALESCE(page.name, '') AS name, COALESCE(page.website, '') AS website, COALESCE(page.notes, '') AS notes,
			COALESCE(page.search_text_lower, '') AS search_text_lower, COALESCE(page.category, '') AS category,
			COALESCE(page.billing_cycle, '') AS billing_cycle, COALESCE(page.currency, '') AS currency,
			COALESCE(page.payment_method, '') AS payment_method, COALESCE(page.status, '') AS status,
			COALESCE(page.pinned, 0) AS pinned, COALESCE(page.public_hidden, 0) AS public_hidden,
			COALESCE(page.next_billing_date, '') AS next_billing_date, COALESCE(page.trial_end_date, '') AS trial_end_date,
			COALESCE(page.one_time_term_count, 0) AS one_time_term_count, COALESCE(page.auto_renew, 0) AS auto_renew,
			COALESCE(page.reminder_days, 0) AS reminder_days,
			COALESCE(page.repeat_reminder_enabled, 0) AS repeat_reminder_enabled,
			COALESCE(page.created_at, '') AS created_at, COALESCE(page.updated_at, '') AS updated_at,
			totals.total_count
			FROM totals LEFT JOIN page ON 1 = 1
			ORDER BY page.created_at DESC, page.subscription_id DESC`, strings.Join(base.conditions, " AND "), strings.Join(pageConditions, " AND "))).
		Bind(params).
		All(&rows)
	return rows, err
}

func appendSQLInCondition(base *subscriptionProjectionBase, column string, prefix string, values []string) {
	if len(values) == 0 {
		return
	}
	placeholders := make([]string, 0, len(values))
	for index, value := range values {
		key := prefix + strconv.Itoa(index)
		placeholders = append(placeholders, "{:"+key+"}")
		base.params[key] = value
	}
	base.conditions = append(base.conditions, column+" IN ("+strings.Join(placeholders, ", ")+")")
}

func appendSQLPaymentMethodCondition(base *subscriptionProjectionBase, values []string) {
	if len(values) == 0 {
		return
	}
	parts := []string{}
	concrete := []string{}
	for _, value := range values {
		if value == subscriptionPaymentMethodNoneValue {
			parts = append(parts, "idx.payment_method = ''")
		} else {
			concrete = append(concrete, value)
		}
	}
	if len(concrete) > 0 {
		placeholders := make([]string, 0, len(concrete))
		for index, value := range concrete {
			key := "paymentMethod" + strconv.Itoa(index)
			placeholders = append(placeholders, "{:"+key+"}")
			base.params[key] = value
		}
		parts = append(parts, "idx.payment_method IN ("+strings.Join(placeholders, ", ")+")")
	}
	base.conditions = append(base.conditions, "("+strings.Join(parts, " OR ")+")")
}

func appendSQLRenewalCondition(base *subscriptionProjectionBase, renewal string) {
	switch renewal {
	case "auto":
		base.conditions = append(base.conditions, "idx.billing_cycle != 'one-time' AND idx.auto_renew = 1")
	case "manual":
		base.conditions = append(base.conditions, "idx.billing_cycle != 'one-time' AND idx.auto_renew = 0")
	case "one-time":
		base.conditions = append(base.conditions, "idx.billing_cycle = 'one-time'")
	}
}

func appendSQLTagCondition(base *subscriptionProjectionBase, values []string) {
	if len(values) == 0 {
		return
	}
	parts := make([]string, 0, len(values))
	for index, tag := range values {
		keyNorm := "tagNorm" + strconv.Itoa(index)
		keyValue := "tag" + strconv.Itoa(index)
		parts = append(parts, "(tag.tag_norm = {:"+keyNorm+"} AND tag.tag = {:"+keyValue+"})")
		base.params[keyNorm] = strings.ToLower(tag)
		base.params[keyValue] = tag
	}
	base.conditions = append(base.conditions, `EXISTS (
		SELECT 1 FROM subscription_tags AS tag
		WHERE tag.user_id = idx.user_id
			AND tag.subscription_id = idx.subscription_id
			AND (`+strings.Join(parts, " OR ")+`)
	)`)
}

func appendSQLReminderModeCondition(base *subscriptionProjectionBase, mode string) {
	switch mode {
	case "disabled":
		base.conditions = append(base.conditions, "idx.reminder_days = -2")
	case "inherit":
		base.conditions = append(base.conditions, "idx.reminder_days = -1")
	case "custom":
		base.conditions = append(base.conditions, "idx.reminder_days >= 0")
	}
}

func subscriptionProjectionIDs(rows []subscriptionListIndexRow) []string {
	ids := make([]string, 0, len(rows))
	for _, row := range rows {
		if row.SubscriptionID != "" {
			ids = append(ids, row.SubscriptionID)
		}
	}
	return ids
}

func subscriptionRecordStringSlice(record *core.Record, name string) []string {
	value := jsonValueForResponse(record.Get(name), []string{})
	switch typed := value.(type) {
	case []string:
		return typed
	case []interface{}:
		out := make([]string, 0, len(typed))
		for _, item := range typed {
			if text, ok := item.(string); ok {
				out = append(out, text)
			}
		}
		return out
	default:
		return []string{}
	}
}

func parseSubscriptionListStrings(values []string, maxItems int, maxLength int, validate func(string) bool) ([]string, error) {
	if len(values) == 0 {
		return nil, nil
	}
	if len(values) > maxItems {
		return nil, errors.New("too many query values")
	}
	seen := make(map[string]struct{}, len(values))
	out := make([]string, 0, len(values))
	for _, raw := range values {
		value := strings.TrimSpace(raw)
		if value == "" || len(value) > maxLength {
			return nil, errors.New("invalid query value")
		}
		if validate != nil && !validate(value) {
			return nil, errors.New("invalid query value")
		}
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		out = append(out, value)
	}
	return out, nil
}

func parseSubscriptionListSingle(values url.Values, name string, maxLength int, validate func(string) bool) (string, error) {
	rawValues := values[name]
	if len(rawValues) == 0 {
		return "", nil
	}
	if len(rawValues) > 1 {
		return "", errors.New("duplicate query value")
	}
	value := strings.TrimSpace(rawValues[0])
	if value == "" || len(value) > maxLength {
		return "", errors.New("invalid query value")
	}
	if validate != nil && !validate(value) {
		return "", errors.New("invalid query value")
	}
	return value, nil
}

func parseSubscriptionListBool(values url.Values, name string) (*bool, error) {
	rawValues := values[name]
	if len(rawValues) == 0 {
		return nil, nil
	}
	if len(rawValues) > 1 {
		return nil, errors.New("duplicate boolean query value")
	}
	var parsed bool
	switch strings.TrimSpace(rawValues[0]) {
	case "true", "1":
		parsed = true
	case "false", "0":
		parsed = false
	default:
		return nil, errors.New("invalid boolean query value")
	}
	return &parsed, nil
}

func isValidSubscriptionListPaymentMethod(value string) bool {
	return value == subscriptionPaymentMethodNoneValue || (strings.TrimSpace(value) == value && value != "")
}

func isSubscriptionListCurrency(value string) bool {
	if len(value) != 3 {
		return false
	}
	for _, char := range value {
		if char < 'A' || char > 'Z' {
			return false
		}
	}
	return true
}

func isSubscriptionListRenewal(value string) bool {
	switch value {
	case "auto", "manual", "one-time":
		return true
	default:
		return false
	}
}

func isSubscriptionListReminderMode(value string) bool {
	switch value {
	case "disabled", "inherit", "custom":
		return true
	default:
		return false
	}
}

File: apps/worker/src/subscription-list-filters.ts
import {
  SUBSCRIPTION_PAYMENT_METHOD_NONE,
  type SubscriptionsListQuery,
} from "@renewlet/shared/schemas/subscriptions";
import { DISABLED_REMINDER_DAYS, INHERIT_REMINDER_DAYS } from "@renewlet/shared/runtime";
import {
  SUBSCRIPTION_COLLECTION_COLUMN_NAMES,
  listSubscriptionCollectionPage,
  parseSubscriptionCursor,
} from "./db";
import { getSubscriptionTotal } from "./subscription-derived-state";
import type { Env, SubscriptionCollectionRow } from "./types";

const SUBSCRIPTION_COLLECTION_COLUMNS_FROM_FACT = SUBSCRIPTION_COLLECTION_COLUMN_NAMES
  .map((column) => `sub.${column}`)
  .join(", ");

export type SubscriptionCollectionFilters = Omit<SubscriptionsListQuery, "cursor" | "limit">;

export interface SubscriptionSqlQueryPlan {
  sql: string;
  params: unknown[];
}

export interface SubscriptionCollectionQueryPlan {
  count: SubscriptionSqlQueryPlan;
  facts: SubscriptionSqlQueryPlan;
}

/**
 * 订阅筛选保留 exact total：筛选条件只作用于轻量投影，事实表 JOIN 也只读取集合 DTO 所需列。
 *
 * cursor 只能影响本页起点，不能进入 total 口径；否则筛选页顶部统计会随滚动递减。
 */
export async function listSubscriptionsForQuery(
  env: Env,
  userId: string,
  query: SubscriptionsListQuery,
  today: string,
): Promise<{ rows: SubscriptionCollectionRow[]; total: number }> {
  if (!subscriptionListQueryHasFilters(query)) {
    // 两个读取互不依赖；并发执行避免无筛选首页形成固定 D1 waterfall。
    const [rows, total] = await Promise.all([
      listSubscriptionCollectionPage(env, userId, { limit: query.limit + 1, cursor: query.cursor }),
      getSubscriptionTotal(env, userId),
    ]);
    return { rows, total };
  }
  return await collectFilteredSubscriptions(env, userId, query, today);
}

/** 完整集合先在投影层确认上限，再读取轻量事实列；超限请求不会触碰 subscriptions facts。 */
export async function listBoundedSubscriptionsForQuery(
  env: Env,
  userId: string,
  query: SubscriptionCollectionFilters,
  today: string,
  maxItems: number,
): Promise<{ rows: SubscriptionCollectionRow[]; total: number; exceeded: boolean }> {
  const plan = subscriptionCollectionQueryPlan(userId, query, today, maxItems + 1);
  const total = await countSubscriptionProjection(env, plan.count);
  if (total > maxItems) return { rows: [], total, exceeded: true };

  // count 与读取之间若发生并发写入，额外一行仍会把请求收敛为 422，不能返回伪完整集合。
  const rows = await readSubscriptionCollectionFacts(env, plan.facts);
  if (rows.length > maxItems) return { rows: [], total: rows.length, exceeded: true };
  // 成功响应以事实行数为 total，避免 count 后并发删除留下与返回数组不一致的元数据。
  return { rows, total: rows.length, exceeded: false };
}

async function collectFilteredSubscriptions(env: Env, userId: string, query: SubscriptionsListQuery, today: string): Promise<{ rows: SubscriptionCollectionRow[]; total: number }> {
  const cursor = parseSubscriptionCursor(query.cursor);
  const plan = subscriptionCollectionQueryPlan(userId, query, today, query.limit + 1, cursor);
  // count 不带业务 cursor，保证 total 描述完整过滤集；事实表 JOIN 避免先收集数千 ID 再构造超大 IN。
  const [total, rows] = await Promise.all([
    countSubscriptionProjection(env, plan.count),
    readSubscriptionCollectionFacts(env, plan.facts),
  ]);
  return { rows, total };
}

type SubscriptionProjectionQuery = { where: string; params: unknown[] };

export function subscriptionCollectionQueryPlan(
  userId: string,
  query: SubscriptionCollectionFilters,
  today: string,
  limit: number,
  cursor?: { createdAt: string; id: string } | null,
): SubscriptionCollectionQueryPlan {
  const base = subscriptionListBaseQuery(userId, query, today);
  const cursorCondition = cursor ? "AND (idx.created_at < ? OR (idx.created_at = ? AND idx.subscription_id < ?))" : "";
  const cursorParams = cursor ? [cursor.createdAt, cursor.createdAt, cursor.id] : [];
  return {
    count: {
      sql: `SELECT COUNT(*) AS total FROM subscription_list_index AS idx WHERE ${base.where}`,
      params: base.params,
    },
    facts: {
      sql: `
        SELECT ${SUBSCRIPTION_COLLECTION_COLUMNS_FROM_FACT}
        FROM subscription_list_index AS idx
        INNER JOIN subscriptions AS sub ON sub.user_id = idx.user_id AND sub.id = idx.subscription_id
        WHERE ${base.where} ${cursorCondition}
        ORDER BY idx.created_at DESC, idx.subscription_id DESC
        LIMIT ?
      `,
      params: [...base.params, ...cursorParams, limit],
    },
  };
}

async function countSubscriptionProjection(env: Env, plan: SubscriptionSqlQueryPlan): Promise<number> {
  const row = await env.DB.prepare(plan.sql)
    .bind(...plan.params)
    .first<{ total: number }>();
  return row?.total ?? 0;
}

async function readSubscriptionCollectionFacts(
  env: Env,
  plan: SubscriptionSqlQueryPlan,
): Promise<SubscriptionCollectionRow[]> {
  const page = await env.DB.prepare(plan.sql).bind(...plan.params).all<SubscriptionCollectionRow>();
  return page.results;
}

function subscriptionListBaseQuery(
  userId: string,
  query: SubscriptionCollectionFilters,
  today: string,
): { where: string; params: unknown[] } {
  // 所有筛选都在 owner-scoped 投影中完成；事实表 JOIN 只负责返回规范化 DTO 所需字段。
  const conditions = ["idx.user_id = ?"];
  const params: unknown[] = [userId];
  appendSqlInCondition(conditions, params, "idx.category", query.category);
  appendSqlInCondition(conditions, params, "idx.billing_cycle", query.billingCycle);
  appendSqlInCondition(conditions, params, "idx.currency", query.currency);
  appendPaymentMethodCondition(conditions, params, query.paymentMethod);
  appendRenewalCondition(conditions, query.renewal);
  appendTagCondition(conditions, params, query.tag);
  if (query.nextBillingFrom) {
    conditions.push("idx.next_billing_date >= ?");
    params.push(query.nextBillingFrom);
  }
  if (query.nextBillingTo) {
    conditions.push("idx.next_billing_date <= ?");
    params.push(query.nextBillingTo);
  }
  if (query.pinned !== undefined) {
    conditions.push("idx.pinned = ?");
    params.push(query.pinned ? 1 : 0);
  }
  if (query.publicHidden !== undefined) {
    conditions.push("idx.public_hidden = ?");
    params.push(query.publicHidden ? 1 : 0);
  }
  appendReminderModeCondition(conditions, params, query.reminderMode);
  if (query.repeatReminder !== undefined) {
    conditions.push("idx.repeat_reminder_enabled = ?");
    params.push(query.repeatReminder ? 1 : 0);
  }
  if (query.q) {
    conditions.push("instr(idx.search_text_lower, ?) > 0");
    params.push(query.q.trim().toLowerCase());
  }
  if (query.status) {
    conditions.push(`(CASE
      WHEN idx.status = 'expired' THEN 'expired'
      WHEN idx.billing_cycle = 'one-time' AND COALESCE(idx.one_time_term_count, 0) <= 0 THEN idx.status
      WHEN idx.status IN ('active', 'trial') AND idx.next_billing_date < ? THEN 'expired'
      ELSE idx.status
    END) = ?`);
    params.push(today, query.status);
  }
  return { where: conditions.join(" AND "), params };
}

function appendSqlInCondition(conditions: string[], params: unknown[], column: string, values: readonly string[] | undefined): void {
  if (!values?.length) return;
  conditions.push(`${column} IN (${values.map(() => "?").join(", ")})`);
  params.push(...values);
}

function appendPaymentMethodCondition(conditions: string[], params: unknown[], values: readonly string[] | undefined): void {
  if (!values?.length) return;
  const concrete = values.filter((value) => value !== SUBSCRIPTION_PAYMENT_METHOD_NONE);
  const parts: string[] = [];
  if (values.includes(SUBSCRIPTION_PAYMENT_METHOD_NONE)) parts.push("(idx.payment_method IS NULL OR idx.payment_method = '')");
  if (concrete.length > 0) {
    parts.push(`idx.payment_method IN (${concrete.map(() => "?").join(", ")})`);
    params.push(...concrete);
  }
  conditions.push(`(${parts.join(" OR ")})`);
}

function appendRenewalCondition(conditions: string[], renewal: SubscriptionsListQuery["renewal"]): void {
  switch (renewal) {
    case "auto":
      conditions.push("idx.billing_cycle != 'one-time' AND idx.auto_renew = 1");
      break;
    case "manual":
      conditions.push("idx.billing_cycle != 'one-time' AND idx.auto_renew = 0");
      break;
    case "one-time":
      conditions.push("idx.billing_cycle = 'one-time'");
      break;
  }
}

function appendTagCondition(conditions: string[], params: unknown[], values: readonly string[] | undefined): void {
  const tags = values?.filter((value) => value.trim() !== "") ?? [];
  if (tags.length === 0) return;
  // tag_norm 用来命中索引，tag 原文用来保留旧 JSON tags.includes 的大小写敏感语义。
  conditions.push(`
    EXISTS (
      SELECT 1 FROM subscription_tags AS tag
      WHERE tag.user_id = idx.user_id
        AND tag.subscription_id = idx.subscription_id
        AND (${tags.map(() => "(tag.tag_norm = ? AND tag.tag = ?)").join(" OR ")})
    )
  `);
  params.push(...tags.flatMap((tag) => [tag.trim().toLowerCase(), tag]));
}

function appendReminderModeCondition(
  conditions: string[],
  params: unknown[],
  mode: SubscriptionsListQuery["reminderMode"],
): void {
  switch (mode) {
    case "disabled":
      conditions.push("idx.reminder_days = ?");
      params.push(DISABLED_REMINDER_DAYS);
      break;
    case "inherit":
      conditions.push("idx.reminder_days = ?");
      params.push(INHERIT_REMINDER_DAYS);
      break;
    case "custom":
      conditions.push("idx.reminder_days >= 0");
      break;
  }
}

function subscriptionListQueryHasFilters(query: SubscriptionCollectionFilters): boolean {
  return Boolean(
    query.q ||
    query.category?.length ||
    query.tag?.length ||
    query.billingCycle?.length ||
    query.paymentMethod?.length ||
    query.currency?.length ||
    query.status ||
    query.renewal ||
    query.nextBillingFrom ||
    query.nextBillingTo ||
    query.pinned !== undefined ||
    query.publicHidden !== undefined ||
    query.reminderMode ||
    query.repeatReminder !== undefined,
  );
}
